miden-lib/asm/scripts/ASWAP.masm

```diff
@@ -0,0 +1,53 @@
+use.miden::sat::account
+use.miden::sat::note
+use.miden::sat::tx
+
+# Atomic Swap script: adds an asset from the note into consumers account and
+# creates a note consumable by initial issuer containing requested ASSET.
+#
+# Requires that the account exposes: 
+#
+# Inputs: []
+# Outputs: []
+#
+# Note inputs are assumed to be as follows:
+#
+# FAILS if:
+begin
+    # Process:
+    # 1. Alice creates a note A containing ASSET X consumable by anyone
+    # 2. Bob consumes note A receiving ASSET X
+    # 3. Bob create a note B containing ASSET Y
+    # 
+    # Requirements:
+    # - Anyone can consume note A
+    # - Note B needs to contain ASSET Y initialy requested by Alice
+    # - Only Alice can consume note B
+
+    # drop the transaction script root
+    dropw
+    # => []
+
+    # load the note inputs to memory starting at address 0
+    push.0 exec.note::get_inputs
+    # => [inputs_ptr] 
+    
+    # load the asset into the stack
+    exec.get_assets drop mem_loadw
+    # => [ASSET]
+
+    # add asset to account
+    exec.account::add_asset dropw
+    # => []
+
+    # TODO: I need more information on the organisation of my memory
+    # and public / private inputs to make sure that I can access and
+    # manipulate them correctly
```

Comment:
I believe after the call to `note::get_inputs`, the inputs will be written to memory in order. So, the first word of inputs will be at `inputs_ptr`, the second word at `inputs_ptr + 1` etc.

With the way you have inputs structured currently, `mem[0]` will contain a word `[faucet_id, amount, tag, recipient]` - where `recipient` is a single field element.

I would recommend structuring inputs as follows: `[RECIPIENT, ASSET, TAG]`, where `TAG = [tag, 0, 0, 0]`, and `RECIPIENT` a word defining the recipient (the way we define it [here](https://github.com/0xPolygonMiden/miden-base/blob/main/objects/src/notes/mod.rs#L166)). This way, memory layout will be as follows:
```
mem[0]: RECIPIENT
mem[1]: ASSET
mem[2]: [tag, 0, 0, 0]
```

This way, you'll be able to read the inputs from memory as follows:
```
padw mem_loadw.0
#=> [RECIPIENT, ...]

padw mem_loadw.1
#=> [ASSET, RECIPIENT, ...]

padw mem_loadw.2
#=> [0, 0, 0, tag, ASSET, RECIPIENT, ...]

drop drop dorp movdn.4
#=> [ASSET, tag, RECIPIENT, ...]
```

Reply:
Implemented following your advice